components/utils.py

```diff
@@ -520,3 +606,282 @@ def untar(source, destination='/tmp', strip=1):
+        '''
+        Archive previously used node properties. These properties will be used
+        for rollback purposes. This method should be called ONLY once it's been
+        determined that the node installation has ended and the service is up.
+        :param service_name: The service/node name.
+        '''
+        service_archive_path = os.path.join(self.ROLLBACK_PROPERTIES_PATH,
+                                            service_name)
+        mkdir(service_archive_path)
+        base_service_dir = os.path.join(self.PROPERTIES_PATH, service_name)
+        properties_file_path = os.path.join(base_service_dir,
+                                            self.PROPERTIES_FILE_NAME)
+        ctx.logger.info('Archiving previous inputs for service {service}'
+                        .format(service=service_name))
+        move(properties_file_path, service_archive_path)
+        ctx.logger.info('Setting new properties for service {service}'
+                        .format(service=service_name))
+        move(self._get_props_file_path(service_name), properties_file_path)
+
+    def _get_props_file_path(self, service_name):
+        base_service_dir = os.path.join(self.PROPERTIES_PATH, service_name)
+        if self.is_upgrade:
+            dest_file_path = os.path.join(
+                    base_service_dir,
+                    'upgrade-{0}'.format(self.PROPERTIES_FILE_NAME))
+        else:
+            dest_file_path = os.path.join(base_service_dir,
+                                          self.PROPERTIES_FILE_NAME)
+        return dest_file_path
+
+    def _load_ctx_properties(self, service_name):
+        node_props = dict(ctx.node.properties.get_all())
+        if self.is_upgrade:
+            # Use existing property configuration during upgrade
+            use_existing = node_props.get('use_existing_on_upgrade')
+            if use_existing:
+                install_props_path = os.path.join(self.PROPERTIES_PATH,
+                                                  service_name,
+                                                  self.PROPERTIES_FILE_NAME)
+                ctx.logger.info('Loading existing {service} input config from '
+                                '{path}'.format(service=service_name,
+                                                path=install_props_path))
+                with open(install_props_path) as f:
+                    existing_props = json.load(f)
+
+                # Removing properties with suffix matching upgrade properties
+                for key in existing_props.keys():
+                    for suffix in self.UPGRADE_PROPS_SUFFIX:
+                        if key.endswith(suffix):
+                            del existing_props[key]
+
+                # Update node properties with existing configuration inputs
+                node_props.update(existing_props)
+
+        node_props['service_name'] = service_name
+
+        return node_props
+
+
```

Comment:
between the two blueprint related classes, I don't see a place in which we backup what's in `/opt/cloudify/sources` which contains the relevant packages for the rollback. Am I missing something?